src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -185,15 +278,18 @@ CHIP_ERROR Register(void * context, DnssdPublishCallback callback, uint32_t inte
         return CHIP_NO_ERROR;
     }
 
-    sdCtx = chip::Platform::New<RegisterContext>(type, callback, context);
-    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
-
-    DNSServiceRef sdRef;
-    auto err = DNSServiceRegister(&sdRef, kRegisterFlags, interfaceId, name, type, kLocalDot, nullptr, ntohs(port), record.size(),
-                                  record.data(), OnRegister, sdCtx);
-    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
+    std::string hostnameStr                   = std::string(hostname) + '.' + kLocalDot;
+    hostname                                  = hostnameStr.c_str();
+    RegisterRecordContext * registerRecordCtx = chip::Platform::New<RegisterRecordContext>(callback, context);
+    VerifyOrReturnError(nullptr != registerRecordCtx, CHIP_ERROR_NO_MEMORY);
+    auto err = RegisterRecord(registerRecordCtx, type, interfaceId, hostname);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, err);
 
-    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+    DNSServiceRef sdRef = nullptr;
+    sdCtx               = chip::Platform::New<RegisterContext>(type, callback, context);
+    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
+    return RegisterService(sdRef, kRegisterFlags, interfaceId, name, type, kLocalDot, hostname, port, record.size(), record.data(),
+                           OnRegister, sdCtx);
```

Comment:
I think that this method needs to be reworked a little bit to be spit in 2 asynchronous phases.

Firstly, `RegisterRecord` is called and only if everything has succeeded then `RegisterService`is done (it means much more fields on `RegisterRecordService` :/)

`RegisterRecord` also needs to track how many `RegisterRecord` calls has been done before calling the next step or even before deleting itself.
For example , if I read correctly the current implementation a `RegisterRecord` on interface `Any` may cause N calls to be emitted. If one fails, then the record will be deleted` and then if we received an other failure, then we will try a double-free.

In the current implementation too, it `RegisterRecord` fails but `RegisterService` succeed, we will likely call both the failure callback and the success callback :)

Reply:
will look into this

Reply:
yeah i thought of that. but was lazy trying not to stuff all the things we need into the context. But i agree its needed